OpenGL context lost/restore handling (pt1)

ekharkunov opened the pull request:
### Technical changes
* Hook context lost/context restore events on JS side.
* Add `render.set_listener` to handle context events (now supported only "context_lost"/"context_restored")
* Pause render if OpenGL context was lost.
* Add array with all GL handles.
* All resources/components now use index from array with GL handles instead of direct storing GL handles. Invalidate all GL handles when context is lost.

Main changes how we deal with invalid GL handles:
Now all GL handles stored in one array `m_AllGLHandles`. When any GL function that generates new handle called - that handle stored in array via `AddNewGLHandle`. Index of stored handle returns outside (for resources/components). When external coed wants to use GL handle - it uses stored index to obtain current GL handle for it (via `GetGLHandle`). When graphics resource was destroyed - called `CleanGLHandle` that set value by index in `m_AllGLHandles` to 0 and stored index in second array `m_FreeIndexes`. `m_FreeIndexes` used to stored unused ("free") indexes in `m_AllGLhandles` array. If any unused index exists in `m_FreeIndexes` than it used for stored new GL handle. If no free index exists - `m_AllGLHandle` grows and stores new handle in the end of array.
When context lost we simply set all stored GL handles to 0.

Relates to #8200

Current use case look like:
1. Set listener in render script via `render.set_listener`.
2. Wait for "context_restored" event.
3. Reboot application.

Script sample (render script):
```lua
function init(self)
    ...........
    pprint("Set render listener")
    render.set_listener(function(_, event_name)
        pprint(event_name)
        if event_name == "context_lost" then
            pprint("context lost")
        elseif event_name == "context_restored" then
            pprint("message posted")
            sys.reboot()
        end
    end)
    pprint("Listener done")
end
```

## PR checklist

* [ ] Code
	* [ ] Add engine and/or editor unit tests.
	* [x] New and changed code follows the overall code style of existing code
	* [x] Add comments where needed
* [ ] Documentation
	* [x] Make sure that API documentation is updated in code comments
	* [ ] Make sure that manuals are updated (in github.com/defold/doc)
* [ ] Prepare pull request and affected issue for automatic release notes generator
	* [ ] Pull request - Write a message that explains what this pull request does. What was the problem? How was it solved? What are the changes to APIs or the new APIs introduced? This message will be used in the generated release notes. Make sure it is well written and understandable for a user of Defold.
	* [ ] Pull request - Write a pull request title that in a sentence summarises what the pull request does. Do not include "Issue-1234 ..." in the title. This text will be used in the generated release notes.
	* [ ] Pull request - Link the pull request to the issue(s) it is closing. Use on of the [approved closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
	* [ ] Affected issue - Assign the issue to a project. Do not assign the pull request to a project if there is an issue which the pull request closes.
	* [ ] Affected issue - Assign the "breaking change" label to the issue if introducing a breaking change.
	* [ ] Affected issue - Assign the "skip release notes" is the issue should not be included in the generated release notes.

----------

Example of a well written PR description:

1. Start with the user facing changes. This will end up in the release notes.
1. Add one of the GitHub approved closing keywords
1. Optionally also add the technical changes made. This is information that might help the reviewer. It will not show up in the release notes. Technical changes are identified by a line starting with one of these:
   1. `### Technical changes` 
   1. `Technical changes:`
   5. `Technical notes:`

```
There was a anomaly in the carbon chroniton propeller, introduced in version 8.10.2. This fix will make sure to reset the phaser collector on application startup.

Fixes #1234

### Technical changes
* Pay special attention to line 23 of phaser_collector.clj as it contains some interesting optimizations
* The propeller code was not taking into account a negative phase.
```

ekharkunov committed:
Hook context lost/context restore events from JS.
Pause render if OpenGL context was lost.
Add array with all GL handles.
All resources/components now use index from array with GL handles instead of direct storing GL handles.
Invalidate all GL handles when context is lost.

ekharkunov commented on engine/gamesys/src/gamesys/resources/res_fragment_program.cpp:
```diff
@@ -18,7 +18,7 @@
 namespace dmGameSystem
 {
-    static dmResource::Result AcquireResources(dmGraphics::HContext context, dmResource::HFactory factory, const char* filename, dmGraphics::ShaderDesc* ddf, dmGraphics::HVertexProgram* program)
+    static dmResource::Result AcquireResources(dmGraphics::HContext context, dmResource::HFactory factory, const char* filename, dmGraphics::ShaderDesc* ddf, dmGraphics::HFragmentProgram* program)
```
Here and below - use right type.

ekharkunov commented on engine/graphics/src/opengl/graphics_opengl.cpp:
```diff
@@ -475,6 +475,58 @@ static void LogFrameBufferError(GLenum status)
 
+    static void UpdateGLHandle(uint32_t idx, GLuint handle)
+    {
+        g_Context->m_AllGLHandles[idx] = handle;
```
Here and below - I use `g_Context` instead of passing `HContext` and cast it to `OpenGLContext` every time. Let me know if such approach is not acceptable.

ekharkunov replied:
I reworked new API to get `HContext` as input parameter. But in the rest of code somewhere I need to use `g_Context` because no context pointer was passed as argument to parent function. If change old API (like `DeleteVertexBuffer`) we need to change lots of place where `gmGraphics::DeleteVertexBuffer` used and pass context as parameter. IMO it's out of scope of the current task. I can do it as a separate set of changes.

ekharkunov commented on engine/render/src/wscript:
```diff
@@ -30,9 +35,15 @@ def build(bld):
       'render/render_script.cpp',
       'render/render_script_camera.cpp'])
+
+    if not waflib.Options.options.skip_build_tests:
```
Guard against flag.

ekharkunov commented on engine/graphics/src/opengl/graphics_opengl_private.h:
```diff
@@ -41,7 +41,7 @@ namespace dmGraphics
         TextureParams     m_Params;
         TextureType       m_Type;
-        GLuint*           m_TextureIds;
+        uint32_t*         m_TextureIds;
```
Here and below - changed type from `GLuint` to `uint32_t`. Instead of storing GL handle directly now store index in `m_AllGLHandles`.

JCash commented on engine/render/src/wscript:
```diff
@@ -10,14 +10,19 @@ def configure(cont):
     pass
 
 def build(bld):
+    sources = bld.path.ant_glob('render/*.cpp') + bld.path.parent.ant_glob('proto/render/*.proto')
```
YEah, setting up dependencies in waf isn't easy.

You _could_ try `bld.add_group()` to manually separate them into groups, that are built one after another. (I.e. group A guaranteed to be finished before group B)

JCash commented on engine/script/src/script/sys_ddf.proto:
```diff
@@ -233,3 +233,18 @@ message SetUpdateFrequency
     required int32 frequency = 1;
 }
+
+/*# resume rendering
+ * Resume rendering.
+ * This message can only be sent to the designated `@system` socket.
+ *
+ * @message
+ * @name resume_rendering
+ * @examples
+ * <pre>
+ * msg.post("@system:", "resume_rendering")
+ * </pre>
+ */
```
Who is supposed to send this and when?
If it's recommended to be sent in the listener, I think it should be mentioned there as well in the example.

JCash reviewed (changes requested):
Overall good imo.
Some smaller implementatin details/improvements.

ekharkunov committed:
Merge remote-tracking branch 'origin/dev' into issue-8200-context-lost

ekharkunov commented on engine/render/src/render/render.cpp:
```diff
@@ -1208,4 +1218,53 @@ namespace dmRender
+    void OnContextEvent(void* context, const char* event_name)
+    {
+        RenderContext* render_context = (RenderContext*)context;
+        if (strcmp(event_name, "context_lost") == 0)
+        {
+            PauseRender(render_context);
+            dmGraphics::InvalidateGraphicsHandles(render_context->m_GraphicsContext);
+        }
```
"context_restored" I will handle in the second part of implementation (will be separate PR). As part of handle it engine will call "reloadResources" and emit "resource_reloaded" event to the listener that user installed via `render.set_listener`.

ekharkunov committed:
Review fixes.
Adopt GlobalVAO to <index, handle> mapping.
Use dmIndexPool instead of dmArray as storage of free GL handles.

ekharkunov committed:
Fixed and simplified condition where check vertex declaration.

ekharkunov commented on engine/render/src/render/render.cpp:
```diff
@@ -1208,4 +1218,53 @@ namespace dmRender
+    void ResumeRender(HRenderContext context)
+    {
+        context->m_IsRenderPaused = 0;
+    }
```
Leave only `SetRenderPause` function.

ekharkunov commented on engine/graphics/src/opengl/graphics_opengl.cpp:
```diff
@@ -1859,14 +1932,14 @@ static void LogFrameBufferError(GLenum status)
+            glBindVertexArray(GetGLHandle(context, context->m_GlobalVAO));
 
-        if (!(context->m_ModificationVersion == vertex_declaration->m_ModificationVersion && vertex_declaration->m_BoundForProgram == ((OpenGLProgram*) program)->m_Id))
+        if (context->m_ModificationVersion != vertex_declaration->m_ModificationVersion || vertex_declaration->m_BoundForProgram != program)
```
Fixed condition and simplified it.
IMO error was in the part `vertex_declaration->m_BoundForProgram == ((OpenGLProgram*) program)->m_Id)`. `m_BoundForProgram` stores `HProgram`.

ekharkunov commented on engine/graphics/src/opengl/graphics_opengl.cpp:
```diff
@@ -1875,7 +1948,7 @@ static void LogFrameBufferError(GLenum status)
                 uint32_t sub_vector_component_count = sub_vector_bind_count > 1 ? sub_vector_bind_count : vertex_declaration->m_Streams[i].m_Size;
 
-                for (int j = 0; j < sub_vector_bind_count; ++j)
+                for (uint32_t j = 0; j < sub_vector_bind_count; ++j)
```
Cosmetic type change. `sub_vector_bind_count` is `uint32_t` type and below functions operates with unsigned types.

ekharkunov committed:
Add constants for event names.

Jhonnyg commented on engine/render/src/wscript:
```diff
@@ -30,9 +35,15 @@ def build(bld):
       'render/render_script_camera.cpp'])
+
+    if not waflib.Options.options.skip_build_tests:
+        bld.recurse('test')
```
newline

Jhonnyg commented on engine/script/src/script/sys_ddf.proto:
```diff
@@ -233,3 +233,18 @@ message SetUpdateFrequency
+ */
+ message ResumeRendering
+ {
+ }
```
newline

Jhonnyg commented on engine/render/src/render/render.cpp:
```diff
@@ -38,6 +38,8 @@ namespace dmRender
     using namespace dmVMath;
 
     const char* RENDER_SOCKET_NAME = "@render";
+    const char* KEY_CONTEXT_LOST = "context_lost";
+    const char* KEY_CONTEXT_RESTORED = "context_restored";
```
Do we need to have these as strings instead of an int/enum?

JCash replied:
I'd prefer int/enum if possible. (Smaller exe size as well as less prone to typos)

ekharkunov replied:
Replace with `dmRender::RenderContextEvent` enumeration. On JS side add 2 field with values for that enumeration. I didn't use `embind` to use constants from C++ because it requires add `-lembind` to linker (and as I understand - add more size to result build).

Jhonnyg reviewed (changes requested):
Generally fine, just some minor cleanups / questions from my side

JCash reviewed (dismissed):
Approved on the conditional that Jhonnys suggestions are also looked into

ekharkunov committed:
Review fixes.

ekharkunov commented:
@Jhonnyg @JCash Please, pay attention to `m_FreeIndexes`. I've changed it back to `dmArray` because index pool implementation doesn't suit very well. Index pool is full in initial state. I need to have an empty pool because it's used as a registry of `free indexes` for other array. So if I use index pool - I need to cleanup it at the beggining and than use. IMO it's a redundant operation here.

ekharkunov committed:
Add missed render library to build_input.yml.
